include/ginkgo/core/base/math.hpp

```diff
@@ -202,6 +202,26 @@ GKO_INLINE GKO_ATTRIBUTES constexpr bool is_complex()
 namespace detail {
 
 
+// singly linked list of all our supported precisions
+template <typename T>
+struct next_precision_impl {};
+
+template <>
+struct next_precision_impl<float> {
+    using type = double;
+};
+
+template <>
+struct next_precision_impl<double> {
+    using type = float;
+};
+
+template <typename T>
+struct next_precision_impl<std::complex<T>> {
+    using type = std::complex<typename next_precision_impl<T>::type>;
+};
```

Comment:
The problem I see with this is that we could no longer use types like `floatX` as `ValueType`, which worked before (I am not 100% sure if we advertised that, but we aimed to support it, see #49 and #64).
Is it possible to add a default dummy-type, so everything would still compile / work?

Alternatively, we could add a helper `supports_precision_change`, which would work in combination with [`std::conditional`](https://en.cppreference.com/w/cpp/types/conditional) for the inheritance and by its own with the `Array` class.

Reply:
We could add new types by adding them to the `next_precision` chain:
Instead of `float -> double -> float`, we could have `float -> double -> half -> bfloat16 -> unum -> double` or add multiple `next_x_precision` for larger steps in the chain (this would allow all pair-wise conversions directly)

Reply:
I don't necessarily want something truly genetic, but I would like a `Dense` matrix to still compile when we use `floatx` as a type instead of our defaults. Obviously, we could not use the precision change feature in that case (which is totally reasonable), but it would be nice if we could use all the other features that do not depend on the precision change.

Reply:
Can you clarify how this would work together with our header/source separation? To me it looks like adding a new precision type requires changes at the implementation level, since it also requires modifying `dense.cpp` or re-implementing all member functions you want to use by yourself.

Reply:
Oh, my bad, I thought we could simply use Ginkgo inside CMake, and use a `floatx` type as a template parameter. After looking into the PRs that I referenced, we simply added the new type to the instantiate macro to make it work. In this case, we would also have to add it to the `next_precision` implementation, but that seems fine to me.
Sorry for causing this much confusion.